ipn/ipnlocal/local.go

```diff
@@ -799,19 +799,22 @@ func (b *LocalBackend) setClientStatus(st controlclient.Status) {
 		b.send(ipn.Notify{LoginFinished: &empty.Message{}})
 	}
 
-	prefsChanged := false
-
 	// Lock b once and do only the things that require locking.
 	b.mu.Lock()
 
 	if st.LogoutFinished != nil {
-		// Since we're logged out now, our netmap cache is invalid.
-		// Since st.NetMap==nil means "netmap is unchanged", there is
-		// no other way to represent this change.
-		b.setNetMapLocked(nil)
-		b.e.SetNetworkMap(new(netmap.NetworkMap))
+		if p := b.pm.CurrentPrefs(); p.Persist() == nil || p.Persist().LoginName == "" {
+			b.mu.Unlock()
+			return
+		}
+		if err := b.pm.DeleteProfile(b.pm.CurrentProfile().ID); err != nil {
+			b.logf("error deleting profile: %v", err)
+		}
+		b.resetForProfileChangeLockedOnEntry()
```

Comment:
Pretty sure that this leads to a deadlock. It manifests itself as a failure when attempting to an OSS roll:

```
--- FAIL: TestIPN (3.48s)
    logger.go:116: 11:54:15.770940 e2e_test.go:180: --- login ---
    logger.go:116: 11:54:15.771051 e2e_test.go:227: n1: not running: NeedsLogin
    logger.go:116: 11:54:15.786566 e2e_test.go:227: n1: not running: Starting
    logger.go:116: 11:54:15.787754 e2e_test.go:227: n1: DERP not connected
    logger.go:116: 11:54:15.890084 e2e_test.go:227: n2: not running: NeedsLogin
    logger.go:116: 11:54:15.906842 e2e_test.go:227: n2: not running: Starting
    logger.go:116: 11:54:15.909070 e2e_test.go:227: n2: DERP not connected
    logger.go:116: 11:54:16.017020 e2e_test.go:291: --- ping n1->n2 ---
    logger.go:116: 11:54:16.120633 e2e_test.go:309: --- logout ---
    logger.go:116: 11:54:16.120658 e2e_test.go:319: n.State=Stopped
    logger.go:116: 11:54:16.121898 e2e_test.go:319: n.State=NoState
    e2e_test.go:324: timeout waiting for logout State=NeedsLogin, got State=NoState
    stuntest.go:64: STUN server shutdown
```

The `LocalClient` is stuck in the `Start` that `resetForProfileChangeLockedOnEntry` tries to do, specifically when we try to shut down the existing controlclient at https://github.com/tailscale/tailscale/blob/1950e56478150728f8ad41436b1cd11afdde5dbe/ipn/ipnlocal/local.go#L1136-L1147. The problem appears to be that we're in a control client state callback already, and we end up waiting for the `c.authDone` channel to be closed, but we're being called from `authRoutine` (which closes it on exit):

```
1 @ 0x100ce8f94 0x100cb39d4 0x100cb3544 0x10116a97c 0x101313938 0x101328a48 0x10131166c 0x101169e84 0x101168550 0x100d1a974
#       0x10116a97b     tailscale.com/control/controlclient.(*Auto).Shutdown+0x1db                              /Users/mihai/Developer/source/tailscale/tailscale/control/controlclient/auto.go:697
#       0x101313937     tailscale.com/ipn/ipnlocal.(*LocalBackend).Start+0x5f7                                  /Users/mihai/Developer/source/tailscale/tailscale/ipn/ipnlocal/local.go:1145
#       0x101328a47     tailscale.com/ipn/ipnlocal.(*LocalBackend).resetForProfileChangeLockedOnEntry+0xa7      /Users/mihai/Developer/source/tailscale/tailscale/ipn/ipnlocal/local.go:4057
#       0x10131166b     tailscale.com/ipn/ipnlocal.(*LocalBackend).setClientStatus+0x51b                        /Users/mihai/Developer/source/tailscale/tailscale/ipn/ipnlocal/local.go:813
#       0x101169e83     tailscale.com/control/controlclient.(*Auto).sendStatus+0x373                            /Users/mihai/Developer/source/tailscale/tailscale/control/controlclient/auto.go:608
#       0x10116854f     tailscale.com/control/controlclient.(*Auto).authRoutine+0x9cf                           /Users/mihai/Developer/source/tailscale/tailscale/control/controlclient/auto.go:322
```

Will poke around, but wanted to flag it in case there's an easy fix.

Reply:
The easiest fix I can think of is to move the call to Start at https://github.com/tailscale/tailscale/blob/1950e56478150728f8ad41436b1cd11afdde5dbe/ipn/ipnlocal/local.go#L4057 into a goroutine, so that it stops leading to reentrancy.

Reply:
I was wondering why this doesn't repro in `TestLogoutRemovesAllPeers` (even with an explicit `AwaitNeedsLogin`) and it's because that uses the `logout` CLI command, which does a sync logout. That's different from the async logout that we do in other tests/code paths.

Reply:
Sent out #6324 as an attempt at fixing this, though it's somewhat of a bandaid.